apollo-router/src/spec/schema.rs

```diff
@@ -22,68 +25,81 @@ use crate::Configuration;
+    diagnostics: Vec<ApolloDiagnostic>,
     subgraphs: HashMap<String, Uri>,
     api_schema: Option<Box<Schema>>,
     pub(crate) schema_id: Option<String>,
 }
 
 #[cfg(test)]
-fn make_api_schema(schema: &str) -> Result<String, SchemaError> {
+fn make_api_schema(schema: &str, configuration: &Configuration) -> Result<String, SchemaError> {
     use itertools::Itertools;
-    use router_bridge::api_schema;
-    let s = api_schema::api_schema(schema)
-        .map_err(|e| SchemaError::Api(e.to_string()))?
-        .map_err(|e| SchemaError::Api(e.iter().filter_map(|e| e.message.as_ref()).join(", ")))?;
+    use router_bridge::api_schema::{api_schema, ApiSchemaOptions};
+    let s = api_schema(
+        schema,
+        ApiSchemaOptions {
+            graphql_validation: matches!(
+                configuration.experimental_graphql_validation,
+                GraphQLValidation::Legacy | GraphQLValidation::Both
+            ),
+        },
+    )
+    .map_err(|e| SchemaError::Api(e.to_string()))?
+    .map_err(|e| SchemaError::Api(e.iter().filter_map(|e| e.message.as_ref()).join(", ")))?;
     Ok(format!("{s}\n"))
 }
 
 impl Schema {
-    pub(crate) fn parse(
-        s: &str,
-        configuration: &Configuration,
-        api_schema: Option<Box<Schema>>,
-    ) -> Result<Self, SchemaError> {
-        let mut schema = Self::parse_inner(s, configuration)?;
-        schema.api_schema = api_schema;
-        Ok(schema)
-    }
-
     #[cfg(test)]
     pub(crate) fn parse_test(s: &str, configuration: &Configuration) -> Result<Self, SchemaError> {
-        let mut schema = Self::parse_inner(s, configuration)?;
-        schema.api_schema = Some(Box::new(Self::parse_inner(
-            &make_api_schema(s)?,
-            configuration,
-        )?));
+        let api_schema = Self::parse(&make_api_schema(s, configuration)?, configuration)?;
+        let schema = Self::parse(s, configuration)?.with_api_schema(api_schema);
         Ok(schema)
     }
 
-    fn parse_inner(schema: &str, _configuration: &Configuration) -> Result<Schema, SchemaError> {
+    pub(crate) fn parse(sdl: &str, configuration: &Configuration) -> Result<Self, SchemaError> {
         let mut compiler = ApolloCompiler::new();
-        let id = compiler.add_type_system(schema, "schema.graphql");
+        let id = compiler.add_type_system(sdl, "schema.graphql");
 
         let ast = compiler.db.ast(id);
 
         // Trace log recursion limit data
         let recursion_limit = ast.recursion_limit();
         tracing::trace!(?recursion_limit, "recursion limit data");
 
-        // TODO: run full compiler-based validation instead?
-        let errors = ast.errors().cloned().collect::<Vec<_>>();
-        if !errors.is_empty() {
-            let errors = ParseErrors {
-                raw_schema: schema.to_string(),
-                errors,
+        let diagnostics =
+            if configuration.experimental_graphql_validation == GraphQLValidation::Legacy {
+                vec![]
```

Comment:
In "legacy" mode, which is the default, this function no longer returns syntax error whereas it previously did. Is this intentional?

Reply:
Was this resolved?